clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -692,23 +692,22 @@ static RValue emitLibraryCall(CodeGenFunction &CGF, const FunctionDecl *FD,
   RValue Call =
       CGF.EmitCall(E->getCallee()->getType(), callee, E, ReturnValueSlot());
 
-  // Check the supported intrinsic.
+  ASTContext &Context = CGF.getContext();
   if (unsigned BuiltinID = FD->getBuiltinID()) {
     auto IsErrnoIntrinsic = [&]() -> unsigned {
-      switch (BuiltinID) {
-      case Builtin::BIexpf:
-      case Builtin::BI__builtin_expf:
-      case Builtin::BI__builtin_expf128:
+      // Check whether a FP math builtin function, such as BI__builtin_expf
+      QualType ResultTy = FD->getReturnType();
+      bool IsMathLibCall =
+          Context.BuiltinInfo.isLibFunction(BuiltinID) ||
+          Context.BuiltinInfo.isPredefinedLibFunction(BuiltinID);
+      if (IsMathLibCall && CGF.ConvertType(ResultTy)->isFloatingPointTy())
```

Comment:
These conditions do not imply it sets errno. Also this can be return of boolean expression

Reply:
My idea was to get the floating-point interface from libc/libm, so I has extra constraints **CGF.ConvertType(ResultTy)->isFloatingPointTy() guarding the return type" and **CGF.CGM.getLangOpts().MathErrno guarding the errno** in the following line 709.

Reply:
Although in the context of this patch, I don't see why any of this is relevant for emitting TBAA

Reply:
For the builtin which doesn't set errno is sure not need add extra TBAA, the origin intend is address these builtins may set errno, such as expf, so we emit "int" TBAA metadata on them will make sure it is safe to reuse the float value from memory.

Reply:
So check isConstWithoutErrnoAndExceptions?

Reply:
No , we expect it works without explicit disable the MathErrno, just similar to the behavior of gcc, https://gcc.godbolt.org/z/rsbqsGWMv. it is discussed on [discourse](https://discourse.llvm.org/t/fp-can-we-add-pure-attribute-for-math-library-functions-default/79459/6?u=allen)

Reply:
Oh, you are right, I see your idea, the expect list can be get by isConstWithoutErrnoAndExceptions